js/private/js_run_devserver.js

```
@@ -0,0 +1,209 @@
+            const numSynced = await syncRecursive(src, dst)
+            totalSynced += numSynced
+        })
+    )
+    var endTime = perf_hooks.performance.now()
+    console.error(
+        `${totalSynced} file${
+            totalSynced > 1 ? 's' : ''
+        } synced in ${Math.round(endTime - startTime)} ms`
+    )
+}
+
+async function main(args, sandbox) {
+    console.error(
+        `\n\nStarting js_run_devserver ${process.env.JS_BINARY__TARGET}`
+    )
+
+    const configPath = path.join(RUNFILES_ROOT, args[0])
+
+    const config = JSON.parse(await fs.promises.readFile(configPath))
+
+    await sync(config.data_files, sandbox)
+
+    return new Promise((resolve) => {
+        const cwd = process.env.JS_BINARY__CHDIR
+            ? path.join(sandbox, process.env.JS_BINARY__CHDIR)
+            : sandbox
+
+        const tool = config.tool
+            ? path.join(RUNFILES_ROOT, config.tool)
+            : config.command
+
+        const toolArgs = args.slice(1)
+
+        console.error(`Running '${tool} ${toolArgs.join(' ')}' in ${cwd}\n\n`)
+
+        const proc = child_process.spawn(tool, toolArgs, {
+            cwd: cwd,
+            stdio: 'inherit',
+            env: {
+                ...process.env,
+                BAZEL_BINDIR: '.',
+                JS_BINARY__CHDIR: '',
+            },
+        })
+
+        proc.on('close', (code) => {
+            console.error(`child tool process exited with code ${code}`)
+            resolve()
+            process.exit(code)
+        })
+
+        process.stdin.on('data', async (chunk) => {
+            const chunkString = chunk.toString()
+            if (chunkString.indexOf('IBAZEL_BUILD_COMPLETED SUCCESS') !== -1) {
+                if (process.env.JS_BINARY__LOG_DEBUG) {
+                    console.error('IBAZEL_BUILD_COMPLETED SUCCESS')
+                }
+                // Re-parse the config file to get the latest list of data files to copy
+                await sync(
```

Do we want to do anything if something within this sync call throws?

---

Just let the process fail with an error message is sufficient

---

Oh I see what you mean. A throw here is an unhandled promise exception.

Fixed.